### PR TITLE
ENH: stats: Add Baumgartner Weiss Schindler test

### PR DESCRIPTION
**Pull request by zhimin-z**
#### Reference issue
Create BWS test for statistical purposes (both naive and variant [Shan + Murakami] methods) (both one-sided and two-sided)

#### What does this implement/fix?
The Baumgartner-Weiss-Schindler (BWS) test is a nonparametric test of the null hypothesis that the distribution underlying sample `x` is the same as the distribution underlying sample `y`. Unlike the Kolmogorv-Smirnov, Wilcoxon, and Cramer-Von Mises tests, the Baumgartner-Weiss-Schindler test weights the integral by the variance of the difference in CDFs, emphasizing the tails of the distributions, which increases the power of the test for a lot of applications.

### CODE REVIEW

**Comment by josef-pkt** [comment]
just one comment, I did not look at the details

IIRC, I read a long time ago that Baumgartner Schindler test has problems when there are ties.
IIRC, the response of one author was to add random noise to the data to break ties.

The argsort argsort will break ties in some numerical way.

So just a question:
Does this work correctly if the user has ties in the data, or, worse case, has many ties when data are discrete?

**Comment by tupui** [comment]
Hi @zhimin-z, thank you for the PR! Before going further we need to get an approval from the mailing list. Could you send an email? https://mail.python.org/mailman3/lists/scipy-dev.python.org/

File: scipy/stats/_bws_test.py
Comment (ilayn) [inline scipy/stats/_bws_test.py:15]: This is going to be very slow. Use numpy instead. 

```suggestion
def calculate_bws_sum(ranks, nm):
    '''Calculate the sum for the Baumgartner-Weiss-Schindler statistic'''
    r1 = ranks + 1
    size = len(ranks)
    factor = 1 / (size + 1)
    i1 = np.arange(1, size + 1)
    ifact = i1 * factor
    return ((r1 - (nm/size)*i1) ** 2 / (ifact*(1-ifact))).sum()
```

**Comment by josef-pkt** [comment]
good, 
(In my zotero I had downloaded those articles in 2013)

The docstring needs to specify which or whether methods are robust to ties.

However, I guess (I don't have the Neuhauser article right now), that it uses midranks for ties. 
Instead of argsort, argsort, you then will need to use the rank function that has midranks used for wilcoxon, ...

**Comment by zhimin-z** [comment]
> 

Done, I think bws is used these days by both the academia and practitioners, as could be witnessed by: https://link.springer.com/article/10.1007/BF02762032
https://pubmed.ncbi.nlm.nih.gov/15284098/
https://www.tandfonline.com/doi/abs/10.1080/03610920008832469
extension://elhekieabhbkpmcefcoobjddigjcaadp/https://cran.r-project.org/web/packages/BWStest/BWStest.pdf

File: scipy/stats/_bws_test.py
Comment (ilayn) [inline scipy/stats/_bws_test.py:31]: There is so much recursion and array resizing here that I can't follow what is happening. It seems like you are recreating the Python built-ins `zip` and `itertools.product`. Why generators? Why concatenate?

**Comment by zhimin-z** [comment]
> Please, before doing any more work (such as doc changes) and review, let's run this through the mailing list. The proposal needs to be strongly motivated as to why we should consider this for inclusion in SciPy.
> 
> See here for a list of things we are looking for https://scipy.github.io/devdocs/dev/core-dev/index.html#deciding-on-new-features
> 
> The main point on my side is: is there a proven track record of this method being widely used and important in some fields? and how does the method compare with our existing methods?

Done, I have sent the email for verification already.
I think bws is used these days by both the academia and practitioners, as could be witnessed by: https://link.springer.com/article/10.1007/BF02762032
https://pubmed.ncbi.nlm.nih.gov/15284098/
https://www.tandfonline.com/doi/abs/10.1080/03610920008832469
https://cran.r-project.org/web/packages/BWStest/BWStest.pdf

File: scipy/stats/_bws_test.py
Comment (ilayn) [inline scipy/stats/_bws_test.py:53]: You are sorting `perm` twice. Sort once then slice but then why are you concatenating again?

File: scipy/stats/_bws_test.py
Comment (ilayn) [inline scipy/stats/_bws_test.py:82]: Create a RNG via `rng = np.random.default_rng()` and then use `rng.shuffle`

https://numpy.org/devdocs/reference/random/generated/numpy.random.Generator.shuffle.html#numpy.random.Generator.shuffle

File: scipy/stats/_bws_test.py
Comment (ilayn) [inline scipy/stats/_bws_test.py:83]: double sorting again

**Comment by tupui** [comment]
Thank you @zhimin-z. I just approved your message.

File: scipy/stats/_bws_test.py
Comment (ilayn) [inline scipy/stats/_bws_test.py:220]: Is this a temporary `pass` or is it going to be finished as such?

**Review by ilayn — CHANGES_REQUESTED** [review]
There is quite some concatenation and sorting happening in these functions. Please write the naive implementation and we can help you speed it up. Most of the computations are spent in array creation and deletion as is.

File: scipy/stats/_bws_test.py
Comment (ilayn) [reply scipy/stats/_bws_test.py:51]: nevermind you are sorting parts individually but then why slicing sort it partially.

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py:218]: I would implement it later according to https://europepmc.org/backend/ptpmcrender.fcgi?accid=PMC4809145&blobtype=pdf

**Comment by josef-pkt** [comment]
I'm all +1 for adding it.
I wanted that test since 2013 :)

(The main disadvantage of permutation tests is that they only allow for restrictive null hypotheses like equality testing, and don't have distribution if equality does not hold.
That's the main reason that permutation tests never got a high priority for me to add to statsmodels.
For example, we can to a lot more besides equality testing with brunner-munzel.
)

**Comment by tupui** [comment]
@ilayn @josef-pkt do I get you are +1 for inclusion and would review/maintain this? If possible I would prefer the team to align first before making any commitment 😃

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py:29]: Hi, @ilayn I adapted from here: https://math.stackexchange.com/questions/4679930/python-realization-of-permutation-of-2-or-more-groups-while-keeping-the-ordering/4679936#4679936

**Comment by ilayn** [comment]
I have no opinion, just went through the code quickly. Also we need more tests 😄

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py:51]: This is adapted from: https://math.stackexchange.com/questions/4679930/python-realization-of-permutation-of-2-or-more-groups-while-keeping-the-ordering/4679936#4679936

**Comment by tupui** [comment]
> I have no opinion, just went through the code quickly. Also we need more tests 😄

lol do we?? 😅 

Jokes aside, anything outstanding you have in mind??

**Comment by josef-pkt** [comment]
I'm just checking the stats and not the implementation details. I'm no expert in permutation tests, nor in the overhead code for scipy.stats function like nan checking.

File: scipy/stats/_bws_test.py
Comment (tupui) [reply scipy/stats/_bws_test.py:29]: I am afraid we cannot do this. This code's license is not compatible with SciPy's license (CC-BY-SA is incompatible with BSD-3-clause). So this part needs to be overhauled.

Comment (zhimin-z) [reply scipy/stats/_bws_test.py:29]: > I am afraid we cannot do this. This code's license is not compatible with SciPy's license (CC-BY-SA is incompatible with BSD-3-clause). So this part needs to be overhauled.

Wait, but this is quite to the original implementation, this is also the question that I initialized. What should we do atm?

Comment (tupui) [reply scipy/stats/_bws_test.py:29]: Just seeing you asked the question yesterday. Then the easy thing that you can try is just ask permission to use in SciPy under our BSD-3-clause. Then in this code add a comment saying the code if from there and we got permission. (There are examples in our codebase you can look at where we did that if you want an example on how to write that.)

Comment (zhimin-z) [reply scipy/stats/_bws_test.py:29]: > Just seeing you asked the question yesterday. Then the easy thing that you can try is just ask permission to use in SciPy under our BSD-3-clause. Then in this code add a comment saying the code if from there and we got permission. (There are examples in our codebase you can look at where we did that if you want an example on how to write that.)

Is there any example for reference?

Comment (tupui) [reply scipy/stats/_bws_test.py:29]:
```
Taken from https://stackoverflow.com/questions/2068372 by P.T. Roy,
written consent given on 23.04.2021 by the original author, Bruno Astrolino,
for free use in SciPy under the 3-clause BSD.
```

Comment (ilayn) [reply scipy/stats/_bws_test.py:29]: I take it back even 

```python
len([x for x in shuffle(np.arange(13).tolist(), np.arange(13).tolist())])  # because it won't work with arrays
```
gives 10 million permutations. Are you sure this is a computable quantity in this way?

Comment (tupui) [reply scipy/stats/_bws_test.py:29]: Just wanted to flag this but even better then 👍 Thanks Ilhan!

Comment (zhimin-z) [reply scipy/stats/_bws_test.py:29]: > I take it back even
> 
> ```python
> len([x for x in shuffle(np.arange(13).tolist(), np.arange(13).tolist())])  # because it won't work with arrays
> ```
> 
> gives 10 million permutations. Are you sure this is a computable quantity in this way?

Thanks. @ilayn @tupui That is the reason why in the original paper, the author does not choose a large array for demonstration. So in this case, I make two modes of estimation: "exact" and "asymptotic", in "exact" mode, we cannot scale to a large array since the permutation number goes crazy. But we can approximate the number by sampling, i.e "asymptotic" mode. In this mode, we can approximate the p-value in a good way without losing too much accuracy when the array goes large.

Comment (rkern) [reply scipy/stats/_bws_test.py:29]: Honestly, I probably would not bother implementing the exact method.

Comment (zhimin-z) [reply scipy/stats/_bws_test.py:29]: > 

Yeah, but it is necessary for verification and benchmarking purpose.

Comment (zhimin-z) [reply scipy/stats/_bws_test.py:29]: > e.g. in `stats._qmc.py`
> 
> ```
> Taken from https://stackoverflow.com/questions/2068372 by P.T. Roy,
> written consent given on 23.04.2021 by the original author, Bruno Astrolino,
> for free use in SciPy under the 3-clause BSD.
> ```

The stack exchange answerer replies to me and we can continue using this @tupui @ilayn, but definitely welcome any optimization!

Comment (ilayn) [reply scipy/stats/_bws_test.py:29]: You can't benchmark anhthing concrete anyways so i would skip exact instead of pretending to have it implemented

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py]: One-sided has been implemented as well @ilayn

**Comment by ilayn** [comment]
@zhimin-z Please take a moment first to assess what we currently have. The exact method is practically not usable. If I accidentally pass arrays more than 13 it waits at least a minute. So that is definitely not going to fly in SciPy regardless of what the theory says. So if there is no approximation or whatever stats equivalent is, we have to find alternatives to implement this. 

Also we have to write quite a number of tests. Without tests you can't assess the correctness of the results while you are developing, and the more commits you are sending the more you are firing up test suites for no reason. Please first finish your rough draft offline then commit here. You can also squash and rebase your commits with meaningful messages.

**Comment by zhimin-z** [comment]
> 

There is a method that turns `exact` off by default, but for test purposes, we need to have an exact implementation, or there won't be any benchmark as far as I found. 

For the test, I am planning to use the exact test on the published papers, it might need some time. @ilayn

**Comment by ilayn** [comment]
You can write the exact version in the tests for comparison.  Please have a look at what I mentioned above regarding pushing commits here. Most of your commit messages are `scrub`.

**Comment by ilayn** [comment]
@zhimin-z Let me put it differently. Everytime you push a commit here for a tiny bit of change this happens 

image


These are virtual machines that are trying to test your version of the code. But you don't have any tests. This is a waste of resources and so far you fired them up 25 times for no reasons. Is this clearer for you? Either add `[ci skip]` to your commit messages so that they don't start. 

Or much better, is that you develop your code on your own laptop until all is ready and then submit it here. So we can review then. Otherwise this is a waste of everybody's time and we are constantly being notified.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [reply scipy/stats/_bws_test.py:6]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.permutation_test.html

Comment (zhimin-z) [reply scipy/stats/_bws_test.py:6]: @mdhaber Nice shot, I did not realize there is ad hoc solution for permutation test, now I have integrated your solution to optimize the current bws test, take a look then!

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:56]: This can be done in a follow-up. Please remove it.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:89]: `method` usually (always) refers to the method used to calculate the p-value. This is more similar to the `variant` parameter of `kendalltau`. Can you use `variant` for the parameter and choose more descriptive strings for the possible inputs?

On the other hand, do we really need both? Are they both commonly used?

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:163]: Needs input validation.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:165]: `permutation_test` accepts `n_resamples`. If it's greater than the number of permutations, the exact test is performed. This should be controlled by the user. The interface depends on how https://github.com/scipy/scipy/issues/18067 is resolved.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:8]: It's ok that you removed `axis` for now. Much slower, but let's keep this as simple as possible.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:165]: Need to calculate ranks before calling the statistic.

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py:89]: The tricky thing is, different authors propose similar `variant` methods, as can be seen in p.5 [Murakami (2006)] and p.7 [Shan (2013)] at https://europepmc.org/backend/ptpmcrender.fcgi?accid=PMC4809145&blobtype=pdf
The way to calculate `B` statistics vary depending on which method you choose.

Comment (rkern) [reply scipy/stats/_bws_test.py:89]: He's suggesting that you name the argument `variant` instead of `method`. That there are multiple variants argues for his suggestion. `method='variant'` does not tell us which variant. `variant='shan'` does. But if you drop `'naive'` like most of us have been suggesting, you don't have to have this argument at all for this first version.

**Comment by tupui** [comment]
Thank you @zhimin-z for using the skip 😃 I see you've figured it out, the correct incantation is `[skip ci]`. In CirrusCI we forgot to add `[ci skip]` this is why it was inconsistent with the rest.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [reply scipy/stats/_bws_test.py]: > Do you mean by _axis_nan_policy_factory or namedtuple? Why not keep it for now?

`_axis_nan_policy_factory` is almost a magic bullet, but not quite. It needs to be configured correctly, and whether it is configured correctly needs to be tested. There will be enough to work on in this PR without that extra complexity, so I am suggesting removing it to make this PR as simple as possible.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:207]: Looks like we still need to calculate ranks.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:124]: A way to control the random state is still needed. Because of that, I'd suggest that we wait on the resolution of gh-18067 before merging this. (I recently reverted my own PR - see gh-18321 - for the same reason.)

**Review by mdhaber — CHANGES_REQUESTED** [review]
If the mailing list thinks it's a good idea to add this, I'll have other comments (separation of `_bws_statistic`, need unit tests), but this is looking much closer to me.

**Comment by tupui** [comment]
What is the `pull bot`?

**Comment by rgommers** [comment]
What on earth is that. I'll look into this asap, because this wasn't me.

**Comment by j-bowhay** [comment]
Looks like Github is getting confused, 419b589 is #18152 not this pr

**Comment by tupui** [comment]
I don't have access but maybe an admin added this app into the orgs or repo apps. In settings I supposed it should show up there
Screenshot 2023-04-24 at 21 51 32

**Comment by rgommers** [comment]
No it's not there. There were only known apps (Azure, Cirrus CI, Travis CI, Codecov, Gitpod, Netlify, Tidelift). Of those, I deleted Codecov and suspended Travis CI and Gitpod. Some of the others didn't have the permissions to merge PRs on this repo. The only two apps left that have these permissions are Azure and Cirrus CI. But they don't seem to be the culprits.

GitHub's audit log says:

image

so I agree with @j-bowhay's assessment that GitHub is confused here. Note the "indirect" in the screenshot above. This is the one I actually merged right before it:

image

I'll keep on looking, but since there is no new commit in `main`, it seems clear that nothing was really merged.

**Comment by rgommers** [comment]
Let me also post a screenshot of what we're seeing right now, in case GitHub unfixes things and we end up even more confused:

image

**Comment by tupui** [comment]
Ok. Now what we also see here is that there is nothing anymore in this PR with 0 commits and no changes.

@zhimin-z or other, did you perform any actions on this branch or your fork?? Do any of you have this installed on their side in some repository? Could be a ripple somehow on GitHub's side

**Comment by tupui** [comment]
I am seeing the following in @zhimin-z's fork

Screenshot 2023-04-24 at 22 24 04

Clicking on one job is pointing to the wheel yaml. https://github.com/zhimin-z/scipy/actions/runs/4788513426

Still looking.

**Comment by tupui** [comment]
So yes this is from @zhimin-z. https://github.com/zhimin-z/scipy/pull/7

**Comment by tupui** [comment]
I think there is a bug due to the fact that the PR was opened on the fork's `main`. I've seen strange GitHub things a few times because of that.

**Comment by wei** [comment]
Yes! Please create a feature branch in the fork and create the PR from the feature branch instead of main. See pull's README for details.

**Comment by wei** [comment]
After Pull executed, there are no changes between the two branches and GitHub automatically closes the PR marking the last commit author as the person who closed the PR via commit.

**Comment by ilayn** [comment]
Probably some github action happened and then deleted, this is a bit too suspicious to delete the code. I'll get some GitHub security people on first. This app shouldn't be able to do any of that.

**Comment by tupui** [comment]
Thanks @ilayn 🙇 also my thinking.

I know GH is closing a PR when the branch disappears, which is close to what is happening as the bot is force pushing the `main` branch where the code was. But the branch is still there, so should not close the PR. And in no circumstances it should show an action as being performed by a maintainer.

**Comment by rgommers** [comment]
> After Pull executed, there are no changes between the two branches and GitHub automatically closes the PR marking the last commit author as the person who closed the PR via commit.

Thank you - that sounds like a reasonable explanation of what happened. I'll verify tomorrow, but I think you may be right. It'd be a GitHub UI bug I'd say, but it's hard to remove all your commits by accident which is why this is a rare event and we're all confused.

**Comment by zhimin-z** [comment]
I created a new feature branch but I found I cannot create a new PR since:
image
What can I do now? @tupui

**Pull request merged**
